lms/djangoapps/progress/signals.py

```diff
@@ -43,3 +50,76 @@ def save_history(sender, instance, **kwargs):  # pylint: disable=no-self-argumen
         completions=instance.completions
     )
     history_entry.save()
+
+
+#
+# Support for Notifications, these two receivers should actually be migrated into a new Leaderboard django app.
+# For now, put the business logic here, but it is pretty decoupled through event signaling
+# so we should be able to move these files easily when we are able to do so
+#
+@receiver(pre_save, sender=StudentProgress)
+def handle_progress_pre_save_signal(sender, instance, **kwargs):
+    """
+    Handle the pre-save ORM event on CourseModuleCompletions
+    """
+
+    if settings.FEATURES['ENABLE_NOTIFICATIONS']:
+        # If notifications feature is enabled, then we need to get the user's
+        # rank before the save is made, so that we can compare it to
+        # after the save and see if the position changes
+        instance.presave_leaderboard_rank = StudentProgress.get_user_position(
+            instance.course_id,
+            instance.user.id,
+            get_aggregate_exclusion_user_ids(instance.course_id)
+        )['position']
```

Comment:
This is a really overloaded call, imho -- I think it should be broken up and protected against a NoneType returned from get_user_position()